whole-tale/ngx-dashboard
# Allow users to share a Tale for active collaboration

## Problem
A Tale can be either `private` or `public`, but this does not allow multiple users to actively collaborate on the same Tale at the same time.

## Approach
Girder offers a set of ACLs for each resource, and we can leverage their API to offer a familiar interface for our users.

Add the Run > Share tab, allowing for more fine-grained management of sharing and permissions.

Demonstration: https://recordit.co/E7JNi9hKwl

Fixes #13 

## How to Test
1. Checkout and run this branch locally, rebuild the dashboard
2. Login to the WholeTale Dashboard
3. Create a Tale, if you haven't already - you don't need to launch/run the Tale
4. Navigate to the Run view for this Tale
    * You should see a new Share tab, indicating the number of collaborators the Tale currently has
5. Navigate to Run > Share
    * You should see radio buttons allowing you to toggle whether the Tale is currently `PRIVATE` or `PUBLIC`
    * You should see the Tale Owner listed here, along with their gravatar
    * You should see a form at the bottom allowing you to invite new collaborators
6. Toggle Tale to `PUBLIC` and refresh the page
    * `PUBLIC` should still be selected after reload
7. Toggle Tale back to `PRIVATE` and refresh the page
    * `PRIVATE` should still be selected after reload
8. Under "Collaborators", click on the user search
    * You should see results appear listing all users
    * NOTE: You will NOT see the Tale owner listed in the collaborators list - removing them from the ACLs gets rid of their access, so we hide them from view to make sure the user can't remove them or modify their access
9. Enter a search query to filter users by name or login, and pick a user
    * You should see the results filter appropriately
    * Upon selecting a user, the "Add" button should be enabled
    * NOTE: You will NOT see the Tale owner listed in the search results - removing them from the ACLs gets rid of their access, so we hide them from view to make sure the user can't remove them or modify their access
10. Choose a user, select "Can Edit" from the dropdown, and click "Add"
    * You should see the user appear in the Collaborators list below
    * You should see that the dropdown beside the user matches the permission you assigned
11. Refresh the page
    * The view should reload exactly the same
12. Toggle the user you added above from "Can Edit" to "Can View" and refresh the page again
    * The user should still have the "Can View" permission, as expected
